back/engines/commercial/impact_tracking/README.md

```diff
@@ -0,0 +1,25 @@
+# Impact Tracking
+
+The impact tracking engine and its corresponding front-end module, track impact by measuring engagement in a privacy-friendly way. These metrics are not exposed in the product, but are used for CitizenLab's internal impact reporting. For this purpose, the numbers need to be highly accurate, but can lack detail.
+
+## Session tracking
+
+Currently the engine only performs session tracking.
+
+### How are sessions tracked?
+
+- Whenever the browser loads the front-end application, on launch, an API request is made to the back-end to store a session
+- The back-end checks whether the user agent is a known crawler, and if so does nothing
+- If the user is human, a session record is stored in the `sessions` table. It stores the current time and a `monthly_user_hash` to uniquely identify the user *within* the current month.
+
+### How is the `monthly_user_hash` calculated?
+
+Since visitors did not accept our cookie policy when the session is stored, we can't rely on a cookie to uniquely identify non-authenticated users that return to the platform. Instead, for non-authenticated visitors, we use the IP and the User-Agent to detect whether a user is returning. This is not fully accurate, but gives a good enough approximation of uniqueness. For authenticated users, we use the user ID.
```

Comment:
Currently only loading the platform triggers a new session, so if they user would do just that, and not revisit, it would be one single session using the unauthenticated hash.

If the user later revisits or refreshes a page, it would add a new session using the authenticated hash. The `monthly_user_hash` would be different between these 2 sessions, so we couldn't identify these sessions as coming from the same person.

I thought about it, but didn't do it (yet?). We could update the `monthly_user_hash` of all the records of the existing month in case the generated visitor_hash for the user matches. It's not impossible but it does complicate the solution somewhat.

Do you think it would be worthwhile?

Reply:
Main reason I didn't do it is performance, it would require every session load for an authenticated user to query the huge sessions table, and also require an index on `monthly_user_hash`. Today, there are 0 reads happening and it's an append- only table, making it all very light and snappy.

That being said (thinking out loud here), we could actually capture the sign-up event and perform the action only then. That wouldn't be too hard indeed, and probably a better solution. I realize now that maybe that's what you had in mind?

Reply:
Yes I think that's what I had in mind - detect the sign-in or signup event, update the events for that user - but then what then happens if they log out again and log in as a different user. Might be better to keep it simple but I do think another event should be captured when they log in as well as on load

Reply:
Ok, I'll do it like that. I'll handle the logic of restarting or updating sessions on the FE to keep that responsibility fully in the FE. It's also easier. Thanks!